File: src/lib/api/collection.ts
import { invoke } from '@tauri-apps/api';

export interface Collection {
	id: string;
	parent_id: string | null;
	name: string;
}

export interface Collections {
	collections: Collection[];
}

export const create = async (name: string, parentId: string | null): Promise<Collection> => {
	const collection = (await invoke('create_collection', { name, parentId })) as Collection;

	return collection;
};

export const list = async (): Promise<Collections> => {
	const collections = (await invoke('list_collections', {})) as Collections;

	return collections;
};

export const listByParentId = async (parentId: string | null): Promise<Collections> => {
	const collections = (await invoke('list_collections_by_parent_id', { parentId })) as Collections;

	return collections;
};

export const get = async (id: string): Promise<Collection> => {
	const collection = (await invoke('get_collection', { id })) as Collection;

	return collection;
};

export const remove = async (id: string): Promise<Collection> => {
	const collection = (await invoke('delete_collection', { id })) as Collection;

	return collection;
};

File: src-tauri/src/cmd/collection.rs
use crate::{
    api::collection::Collections, data::collection::entity, errors::AppError, state::ServiceAccess,
};

#[tauri::command]
pub fn create_collection(
    name: &str,
    parent_id: Option<&str>,
    handle: tauri::AppHandle,
) -> Result<entity::Collection, AppError> {
    handle.collection_ctrl(|ctrl| ctrl.create(name, parent_id))
}

#[tauri::command]
pub fn get_collection(id: &str, handle: tauri::AppHandle) -> Result<entity::Collection, AppError> {
    handle.collection_ctrl(|ctrl| ctrl.get_by_id(id))
}

#[tauri::command]
pub fn update_collection(
    id: &str,
    name: &str,
    parent_id: Option<&str>,
    handle: tauri::AppHandle,
) -> Result<entity::Collection, AppError> {
    handle.collection_ctrl(|ctrl| ctrl.update_by_id(id, name, parent_id))
}

#[tauri::command(async)]
pub fn list_collections(handle: tauri::AppHandle) -> Result<Collections, AppError> {
    handle
        .collection_ctrl(|ctrl| ctrl.list())
        .map(Collections::from)
}

#[tauri::command(async)]
pub fn list_collections_by_parent_id(
    handle: tauri::AppHandle,
    parent_id: Option<String>,
) -> Result<Collections, AppError> {
    handle
        .collection_ctrl(|ctrl| ctrl.list_by_parent_id(parent_id))
        .map(Collections::from)
}

#[tauri::command]
pub fn delete_collection(id: &str, handle: tauri::AppHandle) -> Result<(), AppError> {
    handle.collection_ctrl(|ctrl| ctrl.delete(id))
}
